components/aws/sagemaker/common/_utils.py

```diff
@@ -38,51 +60,102 @@ def get_client(region=None):
 def get_sagemaker_role():
   return get_execution_role()
 
-def create_training_job(client, image, instance_type, instance_count, volume_size, data_location, output_location, role):
+def create_training_job_request(args):
+    with open('/app/common/train.template.yaml', 'r') as f:
+        request = yaml.safe_load(f)
+
+    job_name = args['job_name'] if args['job_name'] else 'TrainingJob-' + strftime("%Y%m%d%H%M%S", gmtime()) + '-' + id_generator()
+
+    request['TrainingJobName'] = job_name
+    request['RoleArn'] = args['role']
+    request['HyperParameters'] = args['hyperparameters']
+    request['AlgorithmSpecification']['TrainingInputMode'] = args['training_input_mode']
+
+    ### Update training image (for BYOC and built-in algorithms) or algorithm resource name
+    if not args['image'] and not args['algorithm_name']:
+        logging.error('Please specify training image or algorithm name.')
+        raise Exception('Could not create job request')
+    if args['image'] and args['algorithm_name']:
+        logging.error('Both image and algorithm name inputted, only one should be specified. Proceeding with image.')
+
+    if args['image']:
+        request['AlgorithmSpecification']['TrainingImage'] = args['image']
+        request['AlgorithmSpecification'].pop('AlgorithmName')
+    else:
+        # TODO: determine if users can make custom algorithm resources that have the same name as built-in algorithm names
```

Comment:
How is custom algorithm supported here?  Ideally, user should use BYOC or first party algorithms. No more options should be exposed

Reply:
Custom algorithm is referring to algorithms that users create as an algorithm resource in SageMaker. The API has an AlgorithmName param for these custom algos and algos from AWS marketplace

Reply:
OK. Have you tried to create a custom algorithm using exist first-party algorithm name? I do think their API object is different or same name can not be used.

Reply:
Just tried it, created an algo with the name 'K-Means' and it worked so I need to adjust this implementation